### PR TITLE
Add some playwright tips to the test strategy document

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description
Will present at retro

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: test-strategy.md [hunk 1]
```diff
@@ -108,3 +108,83 @@ For instructions on how to write and run tests see the [contributing guide](http
...
+![download_trace](https://github.com/gradio-app/gradio/assets/41651716/20c279a8-9a56-4dcf-8df0-c4711e305515)
+
+### Tip 4 - Playwright can write the test for you
+
+You can write the basic skeleton of the test automatically by just interacting with the UI!
+
+First, start a gradio demo from the command line. Then use the following command and point it to the URL of the running demo:
+
+```bash
+npx playwright codegen <gradio-url>
```
Comment (abidlabs) [inline test-strategy.md:183]: just to confirm, this is the regular url, right? might be clearer:
```suggestion
npx playwright codegen <url>
```

File: test-strategy.md [hunk 2]
```diff
@@ -108,3 +108,83 @@ For instructions on how to write and run tests see the [contributing guide](http
...
+For this reason, playright ships with some [retrying assertions](https://playwright.dev/docs/test-assertions#auto-retrying-assertions).
+These assertions will retry until they pass or a timeout is reached, by default 5 seconds.
+So even if playwright checks a DOM element before the server is done, it gives the server a chance to finish by retrying.
+
+
+You can increase the timeout manually as well:
+
+```js
+// 5 seconds
+await expect(page.getByText('Hidden text')).toBeAttached({timeout?: 5000});
```
Comment (abidlabs) [inline test-strategy.md:135]: For clarity, could you include what this test would look like without the retrying assertion?

File: test-strategy.md [hunk 3]
```diff
@@ -108,3 +108,83 @@ For instructions on how to write and run tests see the [contributing guide](http
...
+  const response = await page.request.get('https://api.example.com');
+  expect(response.status()).toBe(200);
+}).toPass();
+```
+
+### Tip 2 - Don't rely on internal network calls to check if something is done
+Internal network calls are not visible to the user, so they can be refactored whenever.
+If we have tests that rely on a request to a given route finishing before moving on, for example, they will fail if we ever change the route name or some other implementation detail.
+
+It's much better to use a retrying assertion that targets a visible DOM element with a larger timeout to check if some work is done.
```
Comment (abidlabs) [inline test-strategy.md:152]: request: could you provide an example of the right vs. wrong thing to do here?

**Comment by abidlabs** [comment]
This is super helpful @freddyaboulton thanks for sharing this! Some small notes above. A nit: can we put these guidelines in a separate document, as this document is more high-level strategy ("...nor does it contain any best practices that we try to follow when writing tests"). We could perhaps create a `testing-guidelines` folder in the repo with these docs, as well as others we might write in the future

**Comment by freddyaboulton** [comment]
I addressed your comments @abidlabs @hannahblair !

**Review by abidlabs — APPROVED** [review]

**Comment by abidlabs** [comment]
Awesome!

**Comment by freddyaboulton** [comment]
Thanks for the reviews!

**Pull request merged**